spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java

```diff
@@ -0,0 +1,189 @@
+                                "Stream name should be non-null, non-empty.");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.awsAccessKeyId),
+                                "Access Key should be non-null, non-empty.");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.awsAccessSecret),
+                                "Access Key secret should be non-null, non-empty.");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.region),
+                                "Region name should be non-null, non-empty.");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.appName),
+                                "Application name should be non-null, non-empty.");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.endpoint),
+                                "Endpoint url should be non-null, non-empty.");
+  }
+
+  @Override
+  public JavaDStream<StructuredRecord> getStream(StreamingContext streamingContext) throws Exception {
+    registerUsage(streamingContext);
+    BasicAWSCredentials awsCred = new BasicAWSCredentials(config.awsAccessKeyId, config.awsAccessSecret);
+    AmazonKinesisClient kinesisClient = new AmazonKinesisClient(awsCred);
+    JavaStreamingContext javaStreamingContext = streamingContext.getSparkStreamingContext();
+    Duration kinesisCheckpointInterval = new Duration(config.duration);
+
+    int numShards = kinesisClient.describeStream(config.streamName).getStreamDescription().getShards().size();
+    List<JavaDStream<byte[]>> streamsList = new ArrayList<>(numShards);
+    LOG.info("creating {} spark executors for {} shards", numShards, numShards);
+    //Creating spark executors based on the number of shards in the stream
+    for (int i = 0; i < numShards; i++) {
+      streamsList.add(
+        KinesisUtils.createStream(javaStreamingContext, config.appName, config.streamName, config.endpoint,
+                                  config.region, config.getInitialPosition(), kinesisCheckpointInterval,
+                                  StorageLevel.MEMORY_AND_DISK_2(), config.awsAccessKeyId, config.awsAccessSecret)
+      );
+    }
+
+    // Union all the streams if there is more than 1 stream
+    JavaDStream<byte[]> kinesisStream;
+    if (streamsList.size() > 1) {
+      kinesisStream = javaStreamingContext.union(streamsList.get(0), streamsList.subList(1, streamsList.size()));
+    } else {
+      // Otherwise, just use the 1 stream
+      kinesisStream = streamsList.get(0);
+    }
+
+    return kinesisStream.map(new Function<byte[], StructuredRecord>() {
+                               public StructuredRecord call(byte[] data) {
+                                 return recordTransform(data);
+                               }
+                             }
+    );
+  }
+
+  private StructuredRecord recordTransform(byte[] data) {
+    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(SCHEMA);
```

Comment:
Why is the SCHEMA just a single string. I remember that we discussed that you should be doing something similar to what KafkaStreamingSource does. If the user knows the schema of the data the source should take the schema and be able to create a structured record through that schema from the byte[].